aws-opentelemetry-distro/tests/amazon/opentelemetry/distro/test_mcpinstrumentor.py

```diff
@@ -0,0 +1,460 @@
+
+    def test_server_handle_request_wrapper_logic(self):
+        """Test the _server_handle_request_wrapper"""
+
+        # Create realistic server request structures
+        class ServerRequest:
+            def __init__(self, has_trace_context=False):
+                if has_trace_context:
+                    self.params = ServerRequestParams(has_meta=True)
+                else:
+                    self.params = ServerRequestParams(has_meta=False)
+
+        class ServerRequestParams:
+            def __init__(self, has_meta=False):
+                if has_meta:
+                    self.meta = ServerRequestMeta()
+                else:
+                    self.meta = None
+
+        class ServerRequestMeta:
+            def __init__(self):
+                self.trace_context = {"trace_id": 12345, "span_id": 67890}
+
+        # Test 1: Request WITHOUT trace context - should take else path
+        request_no_trace = ServerRequest(has_trace_context=False)
+
+        # wrapper's request extraction logic
+        args_with_request = (None, request_no_trace)  # args[1] is the request
+        req = args_with_request[1] if len(args_with_request) > 1 else None
+
+        # Check trace context extraction logic
+        trace_context = None
+        if req and hasattr(req, "params") and req.params and hasattr(req.params, "meta") and req.params.meta:
+            trace_context = req.params.meta.trace_context
+
+        # Verify - should NOT find trace context
+        self.assertIsNotNone(req)
+        self.assertIsNotNone(req.params)
+        self.assertIsNone(req.params.meta)  # No meta field
+        self.assertIsNone(trace_context)
+
+        # Test 2: Request WITH trace context - should take if path
+        request_with_trace = ServerRequest(has_trace_context=True)
+
+        # Simulate the wrapper's request extraction logic
+        args_with_trace = (None, request_with_trace)  # args[1] is the request
+        req2 = args_with_trace[1] if len(args_with_trace) > 1 else None
+
+        # Check trace context extraction logic
+        trace_context2 = None
+        if req2 and hasattr(req2, "params") and req2.params and hasattr(req2.params, "meta") and req2.params.meta:
+            trace_context2 = req2.params.meta.trace_context
+
+        # Verify - should find trace context
+        self.assertIsNotNone(req2)
+        self.assertIsNotNone(req2.params)
+        self.assertIsNotNone(req2.params.meta)  # Has meta field
+        self.assertIsNotNone(trace_context2)
+        self.assertEqual(trace_context2["trace_id"], 12345)
+        self.assertEqual(trace_context2["span_id"], 67890)
+
+        # Test 3: No request at all (args[1] doesn't exist)
+        args_no_request = (None,)  # Only one arg, no request
+        req3 = args_no_request[1] if len(args_no_request) > 1 else None
+
+        # Verify - should handle missing request gracefully
+        self.assertIsNone(req3)
+
+    def test_end_to_end_trace_context_propagation(self):
```

Comment:
IMO context propagation should be tested in contract/integration tests, better to just test basic functionality here.